serialise/src/Codec/Serialise/GeneralisedUTF8.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE BangPatterns #-}
+
+module Codec.Serialise.GeneralisedUTF8 where
+
+import Control.Monad.ST
+import Data.Bits
+import Data.Char
+import Data.Word
+import Data.Primitive.ByteArray
+
+encodeGenUTF8 :: String -> ByteArray
+encodeGenUTF8 st = runST $ do
+    ba <- newByteArray (length st)
+    go ba 0 st
+  where
+    go :: MutableByteArray s -> Int -> [Char] -> ST s ByteArray
+    go ba !_    [] = unsafeFreezeByteArray ba
+    go ba !off  (c:cs)
+      | off + 4 >= cap = do
+        ba' <- newByteArray (cap + cap `div` 2 + 1)
+        copyMutableByteArray ba' 0 ba 0 off
```

Comment:
Oh silly me. I had meant to first shrink the array prior to freezing. I think I was then distracted by adding a shrink primitive to `primitive` and then forgot to use it here.